In [2]:
# !pip list

In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [6]:
import os

In [7]:
os.path.abspath("__file__")

'D:\\E2EML\\E2E-ML\\research\\__file__'

In [8]:
curr_dir = os.path.dirname(os.path.abspath("__file__"))

In [9]:
curr_dir

'D:\\E2EML\\E2E-ML\\research'

In [10]:
matches_path = os.path.join(curr_dir,"..","artifacts","data_ingestion","ipl_dataset","matches.csv")
deliveries_path = os.path.join(curr_dir,"..","artifacts","data_ingestion","ipl_dataset","deliveries.csv")

In [11]:
matches_path

'D:\\E2EML\\E2E-ML\\research\\..\\artifacts\\data_ingestion\\ipl_dataset\\matches.csv'

In [12]:
deliveries_path

'D:\\E2EML\\E2E-ML\\research\\..\\artifacts\\data_ingestion\\ipl_dataset\\deliveries.csv'

In [13]:
matches_data = os.path.normpath(matches_path)
deliveries_data = os.path.normpath(deliveries_path)

In [14]:
print(matches_data)
print(deliveries_data)

D:\E2EML\E2E-ML\artifacts\data_ingestion\ipl_dataset\matches.csv
D:\E2EML\E2E-ML\artifacts\data_ingestion\ipl_dataset\deliveries.csv


In [15]:
match = pd.read_csv(matches_data)
delivery = pd.read_csv(deliveries_data)

In [16]:
match.head()

id    Season       city        date                        team1  \
0   1  IPL-2017  Hyderabad  05-04-2017          Sunrisers Hyderabad   
1   2  IPL-2017       Pune  06-04-2017               Mumbai Indians   
2   3  IPL-2017     Rajkot  07-04-2017                Gujarat Lions   
3   4  IPL-2017     Indore  08-04-2017       Rising Pune Supergiant   
4   5  IPL-2017  Bangalore  08-04-2017  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [17]:
delivery.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [18]:
match.shape

(756, 18)

In [19]:
delivery.shape

(179078, 21)

In [21]:
len(delivery["match_id"].unique())

756

In [95]:
match.columns

Index(['id', 'Season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [96]:
delivery.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

##### All the matches deliveries are present in delivery datset

In [22]:
match.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id               756 non-null    int64 
 1   Season           756 non-null    object
 2   city             749 non-null    object
 3   date             756 non-null    object
 4   team1            756 non-null    object
 5   team2            756 non-null    object
 6   toss_winner      756 non-null    object
 7   toss_decision    756 non-null    object
 8   result           756 non-null    object
 9   dl_applied       756 non-null    int64 
 10  winner           752 non-null    object
 11  win_by_runs      756 non-null    int64 
 12  win_by_wickets   756 non-null    int64 
 13  player_of_match  752 non-null    object
 14  venue            756 non-null    object
 15  umpire1          754 non-null    object
 16  umpire2          754 non-null    object
 17  umpire3          119 non-null    ob

In [23]:
delivery.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

##### Null values are present in both the datasets

In [24]:
delivery[delivery["match_id"]==1]

match_id  inning                 batting_team  \
0           1       1          Sunrisers Hyderabad   
1           1       1          Sunrisers Hyderabad   
2           1       1          Sunrisers Hyderabad   
3           1       1          Sunrisers Hyderabad   
4           1       1          Sunrisers Hyderabad   
..        ...     ...                          ...   
243         1       2  Royal Challengers Bangalore   
244         1       2  Royal Challengers Bangalore   
245         1       2  Royal Challengers Bangalore   
246         1       2  Royal Challengers Bangalore   
247         1       2  Royal Challengers Bangalore   

                    bowling_team  over  ball      batsman  non_striker  \
0    Royal Challengers Bangalore     1     1    DA Warner     S Dhawan   
1    Royal Challengers Bangalore     1     2    DA Warner     S Dhawan   
2    Royal Challengers Bangalore     1     3    DA Warner     S Dhawan   
3    Royal Challengers Bangalore     1     4    DA Warner     S Dhawan   
4    Royal Challengers Bangalore     1     5    DA Warner     S Dhawan   
..                           ...   ...   ...          ...          ...   
243          Sunrisers Hyderabad    19     6    YS Chahal  A Choudhary   
244          Sunrisers Hyderabad    20     1    YS Chahal  A Choudhary   
245          Sunrisers Hyderabad    20     2    YS Chahal  A Choudhary   
246          Sunrisers Hyderabad    20     3  A Choudhary    YS Chahal   
247          Sunrisers Hyderabad    20     4  A Choudhary    YS Chahal   

          bowler  is_super_over  ...  bye_runs  legbye_runs  noball_runs  \
0       TS Mills              0  ...         0            0            0   
1       TS Mills              0  ...         0            0            0   
2       TS Mills              0  ...         0            0            0   
3       TS Mills              0  ...         0            0            0   
4       TS Mills              0  ...         0            0            0   
..           ...            ...  ...       ...          ...          ...   
243      B Kumar              0  ...         0            0            0   
244  BCJ Cutting              0  ...         0            0            0   
245  BCJ Cutting              0  ...         0            0            0   
246  BCJ Cutting              0  ...         0            0            0   
247  BCJ Cutting              0  ...         0            0            0   

     penalty_runs  batsman_runs  extra_runs  total_runs  player_dismissed  \
0               0             0           0           0               NaN   
1               0             0           0           0               NaN   
2               0             4           0           4               NaN   
3               0             0           0           0               NaN   
4               0             0           2           2               NaN   
..            ...           ...         ...         ...               ...   
243             0             1           0           1               NaN   
244             0             0           0           0               NaN   
245             0             1           0           1               NaN   
246             0             6           0           6               NaN   
247             0             0           0           0         YS Chahal   

    dismissal_kind          fielder  
0              NaN              NaN  
1              NaN              NaN  
2              NaN              NaN  
3              NaN              NaN  
4              NaN              NaN  
..             ...              ...  
243            NaN              NaN  
244            NaN              NaN  
245            NaN              NaN  
246            NaN              NaN  
247        run out  CJ Jordan (sub)  

[248 rows x 21 columns]

##### Thus we got to know that match_id 1 is match between SRH & RCB

##### Total of 248 balls played that day

##### Now let's see how much each team has scored?

In [25]:
first_match = delivery[delivery["match_id"]==1]

In [26]:
innings1 = first_match[first_match["inning"]==1]
innings2 = first_match[first_match["inning"]==2]

In [28]:
print(innings1.shape)
print(innings2.shape)

(125, 21)
(123, 21)


In [29]:
innings1.tail()

match_id  inning         batting_team                 bowling_team  over  \
120         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
121         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
122         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
123         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   
124         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore    20   

     ball      batsman  non_striker     bowler  is_super_over  ...  bye_runs  \
120     2     DJ Hooda  BCJ Cutting  SR Watson              0  ...         0   
121     3  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   
122     4  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   
123     5  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   
124     6  BCJ Cutting     DJ Hooda  SR Watson              0  ...         0   

     legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
120            0            0             0             1           0   
121            0            0             0             6           0   
122            0            0             0             2           0   
123            0            0             0             0           0   
124            0            0             0             6           0   

     total_runs  player_dismissed dismissal_kind fielder  
120           1               NaN            NaN     NaN  
121           6               NaN            NaN     NaN  
122           2               NaN            NaN     NaN  
123           0               NaN            NaN     NaN  
124           6               NaN            NaN     NaN  

[5 rows x 21 columns]

##### SRH played all 20 overs

In [30]:
innings2.tail()

match_id  inning                 batting_team         bowling_team  over  \
243         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    19   
244         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   
245         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   
246         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   
247         1       2  Royal Challengers Bangalore  Sunrisers Hyderabad    20   

     ball      batsman  non_striker       bowler  is_super_over  ...  \
243     6    YS Chahal  A Choudhary      B Kumar              0  ...   
244     1    YS Chahal  A Choudhary  BCJ Cutting              0  ...   
245     2    YS Chahal  A Choudhary  BCJ Cutting              0  ...   
246     3  A Choudhary    YS Chahal  BCJ Cutting              0  ...   
247     4  A Choudhary    YS Chahal  BCJ Cutting              0  ...   

     bye_runs  legbye_runs  noball_runs  penalty_runs  batsman_runs  \
243         0            0            0             0             1   
244         0            0            0             0             0   
245         0            0            0             0             1   
246         0            0            0             0             6   
247         0            0            0             0             0   

     extra_runs  total_runs  player_dismissed dismissal_kind          fielder  
243           0           1               NaN            NaN              NaN  
244           0           0               NaN            NaN              NaN  
245           0           1               NaN            NaN              NaN  
246           0           6               NaN            NaN              NaN  
247           0           0         YS Chahal        run out  CJ Jordan (sub)  

[5 rows x 21 columns]

##### RCB only played 19.4 overs

In [41]:
innings1["total_runs"].dtype

dtype('int64')

In [46]:
int(innings1["total_runs"].sum())

207

In [47]:
int(innings2["total_runs"].sum())

172

##### Thus SRH won the match

##### Btw we already knew it from the match dataset

In [48]:
match.head(1)

id    Season       city        date                team1  \
0   1  IPL-2017  Hyderabad  05-04-2017  Sunrisers Hyderabad   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   

   result  dl_applied               winner  win_by_runs  win_by_wickets  \
0  normal           0  Sunrisers Hyderabad           35               0   

  player_of_match                                      venue      umpire1  \
0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal  AY Dandekar   

    umpire2 umpire3  
0  NJ Llong     NaN

In [49]:
match.head(1)["winner"]

0    Sunrisers Hyderabad
Name: winner, dtype: object

##### Now how to get all the matches scores in each innings?

##### Use groupby

In [63]:
delivery.shape

(179078, 21)

In [51]:
pd.DataFrame(delivery.groupby(["match_id","inning"]))

0                                                  1
0         (1, 1)       match_id  inning         batting_team    ...
1         (1, 2)       match_id  inning                 batting_...
2         (2, 1)       match_id  inning    batting_team         ...
3         (2, 2)       match_id  inning            batting_team ...
4         (3, 1)       match_id  inning   batting_team          ...
...          ...                                                ...
1523  (11413, 2)          match_id  inning    batting_team      ...
1524  (11414, 1)          match_id  inning    batting_team      ...
1525  (11414, 2)          match_id  inning         batting_team ...
1526  (11415, 1)          match_id  inning    batting_team      ...
1527  (11415, 2)          match_id  inning         batting_team ...

[1528 rows x 2 columns]

In [64]:
delivery.groupby(["match_id","inning"]).sum()

batting_team  \
match_id inning                                                      
1        1       Sunrisers HyderabadSunrisers HyderabadSunriser...   
         2       Royal Challengers BangaloreRoyal Challengers B...   
2        1       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...   
         2       Rising Pune SupergiantRising Pune SupergiantRi...   
3        1       Gujarat LionsGujarat LionsGujarat LionsGujarat...   
...                                                            ...   
11413    2       Delhi CapitalsDelhi CapitalsDelhi CapitalsDelh...   
11414    1       Delhi CapitalsDelhi CapitalsDelhi CapitalsDelh...   
         2       Chennai Super KingsChennai Super KingsChennai ...   
11415    1       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...   
         2       Chennai Super KingsChennai Super KingsChennai ...   

                                                      bowling_team  over  \
match_id inning                                                            
1        1       Royal Challengers BangaloreRoyal Challengers B...  1312   
         2       Sunrisers HyderabadSunrisers HyderabadSunriser...  1259   
2        1       Rising Pune SupergiantRising Pune SupergiantRi...  1323   
         2       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...  1270   
3        1       Kolkata Knight RidersKolkata Knight RidersKolk...  1280   
...                                                            ...   ...   
11413    2       Sunrisers HyderabadSunrisers HyderabadSunriser...  1301   
11414    1       Chennai Super KingsChennai Super KingsChennai ...  1294   
         2       Delhi CapitalsDelhi CapitalsDelhi CapitalsDelh...  1160   
11415    1       Chennai Super KingsChennai Super KingsChennai ...  1287   
         2       Mumbai IndiansMumbai IndiansMumbai IndiansMumb...  1295   

                 ball                                            batsman  \
match_id inning                                                            
1        1        456  DA WarnerDA WarnerDA WarnerDA WarnerDA WarnerS...   
         2        444  CH GayleMandeep SinghMandeep SinghMandeep Sing...   
2        1        455  PA PatelPA PatelPA PatelJC ButtlerPA PatelPA P...   
         2        435  AM RahaneMA AgarwalMA AgarwalMA AgarwalAM Raha...   
3        1        434  JJ RoyJJ RoyJJ RoyJJ RoyJJ RoyJJ RoyBB McCullu...   
...               ...                                                ...   
11413    2        448  P ShawP ShawS DhawanS DhawanS DhawanP ShawS Dh...   
11414    1        441  P ShawP ShawP ShawP ShawP ShawS DhawanP ShawS ...   
         2        413  F du PlessisF du PlessisF du PlessisSR WatsonS...   
11415    1        447  Q de KockRG SharmaQ de KockQ de KockQ de KockQ...   
         2        448  F du PlessisF du PlessisF du PlessisF du Pless...   

                                                       non_striker  \
match_id inning                                                      
1        1       S DhawanS DhawanS DhawanS DhawanS DhawanDA War...   
         2       Mandeep SinghCH GayleCH GayleCH GayleCH GayleC...   
2        1       JC ButtlerJC ButtlerJC ButtlerPA PatelJC Buttl...   
         2       MA AgarwalAM RahaneAM RahaneAM RahaneMA Agarwa...   
3        1       BB McCullumBB McCullumBB McCullumBB McCullumBB...   
...                                                            ...   
11413    2       S DhawanS DhawanP ShawP ShawP ShawS DhawanP Sh...   
11414    1       S DhawanS DhawanS DhawanS DhawanS DhawanP Shaw...   
         2       SR WatsonSR WatsonSR WatsonF du PlessisF du Pl...   
11415    1       RG SharmaQ de KockRG SharmaRG SharmaRG SharmaR...   
         2       SR WatsonSR WatsonSR WatsonSR WatsonF du Pless...   

                                                            bowler  \
match_id inning                                                      
1        1       TS MillsTS MillsTS MillsTS MillsTS MillsTS Mil...   
         2       A NehraA NehraA NehraA Nehr

In [53]:
delivery.groupby(["match_id","inning"]).sum()["total_runs"]

match_id  inning
1         1         207
          2         172
2         1         184
          2         187
3         1         183
                   ... 
11413     2         170
11414     1         155
          2         162
11415     1         152
          2         157
Name: total_runs, Length: 1528, dtype: int64

In [69]:
pd.DataFrame(delivery[delivery["inning"]==1].groupby(["match_id"]))

0                                                  1
0        (1,)       match_id  inning         batting_team    ...
1        (2,)       match_id  inning    batting_team         ...
2        (3,)       match_id  inning   batting_team          ...
3        (4,)       match_id  inning            batting_team ...
4        (5,)        match_id  inning                 batting...
..        ...                                                ...
751  (11347,)          match_id  inning           batting_tea...
752  (11412,)          match_id  inning         batting_team ...
753  (11413,)          match_id  inning         batting_team ...
754  (11414,)          match_id  inning    batting_team      ...
755  (11415,)          match_id  inning    batting_team      ...

[756 rows x 2 columns]

In [71]:
pd.DataFrame(delivery[delivery["inning"]==2].groupby(["match_id"]))

0                                                  1
0        (1,)       match_id  inning                 batting_...
1        (2,)       match_id  inning            batting_team ...
2        (3,)       match_id  inning           batting_team  ...
3        (4,)       match_id  inning     batting_team        ...
4        (5,)        match_id  inning      batting_team      ...
..        ...                                                ...
749  (11347,)          match_id  inning    batting_team      ...
750  (11412,)          match_id  inning    batting_team      ...
751  (11413,)          match_id  inning    batting_team      ...
752  (11414,)          match_id  inning         batting_team ...
753  (11415,)          match_id  inning         batting_team ...

[754 rows x 2 columns]

In [70]:
match.shape

(756, 18)

##### That means there are no second innings for 2 matches

In [73]:
match["dl_applied"].value_counts()

dl_applied
0    737
1     19
Name: count, dtype: int64

In [74]:
match["result"].value_counts()

result
normal       743
tie            9
no result      4
Name: count, dtype: int64

In [75]:
match[match["result"]=="no result"]

id    Season       city        date                        team1  \
300    301  IPL-2011      Delhi  21-05-2011             Delhi Daredevils   
545    546  IPL-2015  Bangalore  29-04-2015  Royal Challengers Bangalore   
570    571  IPL-2015  Bangalore  17-05-2015             Delhi Daredevils   
744  11340  IPL-2019  Bengaluru  30-04-2019  Royal Challengers Bangalore   

                           team2                  toss_winner toss_decision  \
300                Pune Warriors             Delhi Daredevils           bat   
545             Rajasthan Royals             Rajasthan Royals         field   
570  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
744             Rajasthan Royals             Rajasthan Royals         field   

        result  dl_applied winner  win_by_runs  win_by_wickets  \
300  no result           0    NaN            0               0   
545  no result           0    NaN            0               0   
570  no result           0    NaN            0               0   
744  no result           0    NaN            0               0   

    player_of_match                   venue          umpire1       umpire2  \
300             NaN        Feroz Shah Kotla        SS Hazare     RJ Tucker   
545             NaN   M Chinnaswamy Stadium        JD Cloete     PG Pathak   
570             NaN   M Chinnaswamy Stadium  HDPK Dharmasena  K Srinivasan   
744             NaN  M. Chinnaswamy Stadium      Nigel Llong  Ulhas Gandhe   

            umpire3  
300             NaN  
545             NaN  
570             NaN  
744  Anil Chaudhary

In [79]:
noResultMatchIds = match[match["result"]=="no result"]["id"].values
noResultMatchIds

array([  301,   546,   571, 11340])

In [85]:
noResultDeliveries = delivery[delivery["match_id"].isin(noResultMatchIds)]

In [88]:
noResultDeliveries.groupby(["match_id","inning"]).sum()["total_runs"]

match_id  inning
301       1          56
546       1         200
571       1         187
          2           2
11340     1          65
          2          42
Name: total_runs, dtype: int64

##### Thus match_id 301 & 546 don't have any second innings

In [113]:
delivery.groupby("match_id")["inning"]

In [115]:
singleInningMatches = delivery.groupby("match_id")["inning"].apply(lambda x:2 not in x.values)

In [118]:
singleInningMatches

match_id
1        False
2        False
3        False
4        False
5        False
         ...  
11347    False
11412    False
11413    False
11414    False
11415    False
Name: inning, Length: 756, dtype: bool

In [129]:
singleInnings = np.array(singleInningMatches[singleInningMatches.values==True].index)

In [142]:
singleInnings

array([301, 546])

In [130]:
match[match["id"].isin(singleInnings)]

id    Season       city        date                        team1  \
300  301  IPL-2011      Delhi  21-05-2011             Delhi Daredevils   
545  546  IPL-2015  Bangalore  29-04-2015  Royal Challengers Bangalore   

                team2       toss_winner toss_decision     result  dl_applied  \
300     Pune Warriors  Delhi Daredevils           bat  no result           0   
545  Rajasthan Royals  Rajasthan Royals         field  no result           0   

    winner  win_by_runs  win_by_wickets player_of_match  \
300    NaN            0               0             NaN   
545    NaN            0               0             NaN   

                     venue    umpire1    umpire2 umpire3  
300       Feroz Shah Kotla  SS Hazare  RJ Tucker     NaN  
545  M Chinnaswamy Stadium  JD Cloete  PG Pathak     NaN

##### Now let's try to merge match df and delivery df

As observed, match df doesn't contain any info regarding total runs, so let's add 1st inning total runs from delivery df to match df

In [131]:
delivery.groupby(["match_id","inning"]).sum()["total_runs"]

match_id  inning
1         1         207
          2         172
2         1         184
          2         187
3         1         183
                   ... 
11413     2         170
11414     1         155
          2         162
11415     1         152
          2         157
Name: total_runs, Length: 1528, dtype: int64

In [132]:
delivery.groupby(["match_id","inning"]).sum()["total_runs"].reset_index()

match_id  inning  total_runs
0            1       1         207
1            1       2         172
2            2       1         184
3            2       2         187
4            3       1         183
...        ...     ...         ...
1523     11413       2         170
1524     11414       1         155
1525     11414       2         162
1526     11415       1         152
1527     11415       2         157

[1528 rows x 3 columns]

In [133]:
total_scores = delivery.groupby(["match_id","inning"]).sum()["total_runs"].reset_index()

In [134]:
total_score_innings1 = total_scores[total_scores["inning"]==1]

In [135]:
total_score_innings1.head()

match_id  inning  total_runs
0         1       1         207
2         2       1         184
4         3       1         183
6         4       1         163
8         5       1         157

In [136]:
total_score_innings1 = total_score_innings1.drop(["inning"],axis=1)

In [139]:
total_score_innings1.head(2)

match_id  total_runs
0         1         207
2         2         184

##### Merge 2 dfs

In [271]:
match_totalRuns1 = match.merge(total_score_innings1, left_on="id", right_on="match_id")
match_totalRuns1.head()

id    Season       city        date                        team1  \
0   1  IPL-2017  Hyderabad  05-04-2017          Sunrisers Hyderabad   
1   2  IPL-2017       Pune  06-04-2017               Mumbai Indians   
2   3  IPL-2017     Rajkot  07-04-2017                Gujarat Lions   
3   4  IPL-2017     Indore  08-04-2017       Rising Pune Supergiant   
4   5  IPL-2017  Bangalore  08-04-2017  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  match_id  total_runs  
0     AY Dandekar       NJ Llong     NaN         1         207  
1  A Nand Kishore         S Ravi     NaN         2         184  
2     Nitin Menon      CK Nandan     NaN         3         183  
3    AK Chaudhary  C Shamshuddin     NaN         4         163  
4             NaN            NaN     NaN         5         157

In [272]:
match_totalRuns1.shape

(756, 20)

In [273]:
match.shape

(756, 18)

##### Match id's without 2nd innings will still have 1st innings total scores

In [274]:
match_totalRuns1[match_totalRuns1["id"].isin(singleInnings)]

id    Season       city        date                        team1  \
300  301  IPL-2011      Delhi  21-05-2011             Delhi Daredevils   
545  546  IPL-2015  Bangalore  29-04-2015  Royal Challengers Bangalore   

                team2       toss_winner toss_decision     result  dl_applied  \
300     Pune Warriors  Delhi Daredevils           bat  no result           0   
545  Rajasthan Royals  Rajasthan Royals         field  no result           0   

    winner  win_by_runs  win_by_wickets player_of_match  \
300    NaN            0               0             NaN   
545    NaN            0               0             NaN   

                     venue    umpire1    umpire2 umpire3  match_id  total_runs  
300       Feroz Shah Kotla  SS Hazare  RJ Tucker     NaN       301          56  
545  M Chinnaswamy Stadium  JD Cloete  PG Pathak     NaN       546         200

##### Let's check the team names

In [275]:
match_totalRuns1["team1"].unique()

array(['Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Royal Challengers Bangalore',
       'Kolkata Knight Riders', 'Delhi Daredevils', 'Kings XI Punjab',
       'Chennai Super Kings', 'Rajasthan Royals', 'Deccan Chargers',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

In [276]:
match_totalRuns1["team2"].unique()

array(['Royal Challengers Bangalore', 'Rising Pune Supergiant',
       'Kolkata Knight Riders', 'Kings XI Punjab', 'Delhi Daredevils',
       'Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Lions',
       'Rajasthan Royals', 'Chennai Super Kings', 'Deccan Chargers',
       'Pune Warriors', 'Kochi Tuskers Kerala', 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

In [278]:
match_totalRuns1["toss_winner"].unique()

array(['Royal Challengers Bangalore', 'Rising Pune Supergiant',
       'Kolkata Knight Riders', 'Kings XI Punjab', 'Sunrisers Hyderabad',
       'Mumbai Indians', 'Gujarat Lions', 'Delhi Daredevils',
       'Chennai Super Kings', 'Rajasthan Royals', 'Deccan Chargers',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

In [279]:
match_totalRuns1["winner"].unique()

array(['Sunrisers Hyderabad', 'Rising Pune Supergiant',
       'Kolkata Knight Riders', 'Kings XI Punjab',
       'Royal Challengers Bangalore', 'Mumbai Indians',
       'Delhi Daredevils', 'Gujarat Lions', 'Chennai Super Kings',
       'Rajasthan Royals', 'Deccan Chargers', 'Pune Warriors',
       'Kochi Tuskers Kerala', nan, 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

In [280]:
currentTeamNames = [
    "Chennai Super Kings",
    "Delhi Capitals",
    "Gujarat Titans",
    "Kolkata Knight Riders",
    "Lucknow Super Giants",
    "Mumbai Indians",
    "Punjab Kings",
    "Rajasthan Royals",
    "Royal Challengers Bangalore,"
    "Sunrisers Hyderabad"
]

In [281]:
def updateTeamNames(df,oldTeamName,newTeamName):
    df["team1"] = df["team1"].str.replace(oldTeamName, newTeamName)
    df["team2"] = df["team2"].str.replace(oldTeamName, newTeamName)
    df["toss_winner"] = df["toss_winner"].str.replace(oldTeamName, newTeamName)
    df["winner"] = df["winner"].str.replace(oldTeamName, newTeamName)

In [282]:
updateTeamNames(match_totalRuns1, "Deccan Chargers", "Sunrisers Hyderabad")
updateTeamNames(match_totalRuns1, "Gujarat Lions", "Gujarat Titans")
updateTeamNames(match_totalRuns1, "Rising Pune Supergiant", "Chennai Super Kings")
updateTeamNames(match_totalRuns1, "Chennai Super Kingss", "Chennai Super Kings")
updateTeamNames(match_totalRuns1, "Delhi Daredevils", "Delhi Capitals")
updateTeamNames(match_totalRuns1, "Kings XI Punjab", "Punjab Kings")

In [256]:
match_totalRuns1.shape

(756, 20)

In [283]:
match_totalRuns1["team1"].value_counts()

team1
Sunrisers Hyderabad            106
Chennai Super Kings            104
Mumbai Indians                 101
Punjab Kings                    91
Royal Challengers Bangalore     85
Kolkata Knight Riders           83
Delhi Capitals                  78
Rajasthan Royals                67
Pune Warriors                   20
Gujarat Titans                  14
Kochi Tuskers Kerala             7
Name: count, dtype: int64

In [284]:
match_totalRuns1["team2"].value_counts()

team2
Delhi Capitals                 99
Kolkata Knight Riders          95
Royal Challengers Bangalore    95
Chennai Super Kings            90
Mumbai Indians                 86
Punjab Kings                   85
Rajasthan Royals               80
Sunrisers Hyderabad            77
Pune Warriors                  26
Gujarat Titans                 16
Kochi Tuskers Kerala            7
Name: count, dtype: int64

In [259]:
match_totalRuns1.shape

(756, 20)

In [285]:
newdf1=match_totalRuns1[match_totalRuns1["team1"].isin(currentTeamNames)]

In [286]:
newdf1.shape

(538, 20)

In [287]:
newdf2=match_totalRuns1[match_totalRuns1["team2"].isin(currentTeamNames)]

In [288]:
newdf2.shape

(551, 20)

In [289]:
newdf2["team1"].value_counts()

team1
Sunrisers Hyderabad            83
Royal Challengers Bangalore    73
Mumbai Indians                 72
Punjab Kings                   71
Chennai Super Kings            71
Rajasthan Royals               51
Delhi Capitals                 50
Kolkata Knight Riders          50
Pune Warriors                  18
Gujarat Titans                  8
Kochi Tuskers Kerala            4
Name: count, dtype: int64

In [290]:
newdf2["team2"].value_counts()

team2
Delhi Capitals           99
Kolkata Knight Riders    95
Chennai Super Kings      90
Mumbai Indians           86
Punjab Kings             85
Rajasthan Royals         80
Gujarat Titans           16
Name: count, dtype: int64

In [294]:
newdf2[newdf2["team1"].isin(currentTeamNames)].shape

(373, 20)

In [247]:
match_totalRuns1[match_totalRuns1["team2"].isin(currentTeamNames)]

id    Season           city        date                  team1  \
1        2  IPL-2017           Pune  06-04-2017         Mumbai Indians   
2        3  IPL-2017         Rajkot  07-04-2017         Gujarat Titans   
3        4  IPL-2017         Indore  08-04-2017    Chennai Super Kings   
6        7  IPL-2017         Mumbai  09-04-2017  Kolkata Knight Riders   
8        9  IPL-2017           Pune  11-04-2017         Delhi Capitals   
..     ...       ...            ...         ...                    ...   
750  11346  IPL-2019         Mohali  05-05-2019    Chennai Super Kings   
751  11347  IPL-2019         Mumbai  05-05-2019  Kolkata Knight Riders   
752  11412  IPL-2019        Chennai  07-05-2019    Chennai Super Kings   
754  11414  IPL-2019  Visakhapatnam  10-05-2019         Delhi Capitals   
755  11415  IPL-2019      Hyderabad  12-05-2019         Mumbai Indians   

                     team2            toss_winner toss_decision  result  \
1      Chennai Super Kings    Chennai Super Kings         field  normal   
2    Kolkata Knight Riders  Kolkata Knight Riders         field  normal   
3             Punjab Kings           Punjab Kings         field  normal   
6           Mumbai Indians         Mumbai Indians         field  normal   
8      Chennai Super Kings    Chennai Super Kings         field  normal   
..                     ...                    ...           ...     ...   
750           Punjab Kings           Punjab Kings         field  normal   
751         Mumbai Indians         Mumbai Indians         field  normal   
752         Mumbai Indians    Chennai Super Kings           bat  normal   
754    Chennai Super Kings    Chennai Super Kings         field  normal   
755    Chennai Super Kings         Mumbai Indians           bat  normal   

     dl_applied                 winner  win_by_runs  win_by_wickets  \
1             0    Chennai Super Kings            0               7   
2             0  Kolkata Knight Riders            0              10   
3             0           Punjab Kings            0               6   
6             0         Mumbai Indians            0               4   
8             0         Delhi Capitals           97               0   
..          ...                    ...          ...             ...   
750           0           Punjab Kings            0               6   
751           0         Mumbai Indians            0               9   
752           0         Mumbai Indians            0               6   
754           0    Chennai Super Kings            0               6   
755           0         Mumbai Indians            1               0   

    player_of_match                                    venue  \
1         SPD Smith  Maharashtra Cricket Association Stadium   
2           CA Lynn   Saurashtra Cricket Association Stadium   
3        GJ Maxwell                   Holkar Cricket Stadium   
6            N Rana                         Wankhede Stadium   
8         SV Samson  Maharashtra Cricket Association Stadium   
..              ...                                      ...   
750        KL Rahul                        IS Bindra Stadium   
751       HH Pandya                         Wankhede Stadium   
752        AS Yadav                M. A. Chidambaram Stadium   
754    F du Plessis                         ACA-VDCA Stadium   
755       JJ Bumrah       Rajiv Gandhi Intl. Cricket Stadium   

                   umpire1         umpire2                  umpire3  match_id  \
1           A Nand Kishore          S Ravi                      NaN         2   
2              Nitin Menon       CK Nandan                      NaN         3   
3             AK Chaudhary   C Shamshuddin                      NaN         4   
6              Nitin Menon       CK Nandan                      NaN         7   
8              AY Dandekar          S Ravi                      NaN         9   
..                     ...             ...                      ...       ...   
750  KN Ananthapadmanabhan   

In [225]:
match_totalRuns2 = match_totalRuns1[match_totalRuns1["team2"].isin(currentTeamNames)]
match_totalRuns2.shape

(363, 20)

In [209]:
matchesLatestTeams=onlyIncludeCurrentTeams(match_totalRuns1, currentTeamNames)

In [210]:
matchesLatestTeams.shape

(363, 20)

In [192]:
matchesLatestTeams["team1"].unique()

array(['Mumbai Indians', 'Gujarat Titans', 'Chennai Super Kings',
       'Royal Challengers Bangalore', 'Kolkata Knight Riders',
       'Delhi Capitals', 'Sunrisers Hyderabad', 'Punjab Kings',
       'Rajasthan Royals', 'Pune Warriors', 'Kochi Tuskers Kerala',
       'Chennai Super Kingss'], dtype=object)

In [193]:
matchesLatestTeams["team2"].unique()

array(['Chennai Super Kings', 'Kolkata Knight Riders', 'Punjab Kings',
       'Delhi Capitals', 'Sunrisers Hyderabad', 'Mumbai Indians',
       'Gujarat Titans', 'Royal Challengers Bangalore',
       'Rajasthan Royals', 'Pune Warriors', 'Kochi Tuskers Kerala',
       'Chennai Super Kingss'], dtype=object)

In [161]:
matchesLatestTeams.shape

(363, 20)

In [163]:
match.shape

(756, 18)

In [164]:
matchesLatestTeams.head()

id    Season    city        date                  team1  \
1   2  IPL-2017    Pune  06-04-2017         Mumbai Indians   
2   3  IPL-2017  Rajkot  07-04-2017         Gujarat Titans   
3   4  IPL-2017  Indore  08-04-2017    Chennai Super Kings   
6   7  IPL-2017  Mumbai  09-04-2017  Kolkata Knight Riders   
8   9  IPL-2017    Pune  11-04-2017         Delhi Capitals   

                   team2             toss_winner toss_decision  result  \
1    Chennai Super Kings  Rising Pune Supergiant         field  normal   
2  Kolkata Knight Riders   Kolkata Knight Riders         field  normal   
3           Punjab Kings         Kings XI Punjab         field  normal   
6         Mumbai Indians          Mumbai Indians         field  normal   
8    Chennai Super Kings  Rising Pune Supergiant         field  normal   

   dl_applied                  winner  win_by_runs  win_by_wickets  \
1           0  Rising Pune Supergiant            0               7   
2           0   Kolkata Knight Riders            0              10   
3           0         Kings XI Punjab            0               6   
6           0          Mumbai Indians            0               4   
8           0        Delhi Daredevils           97               0   

  player_of_match                                    venue         umpire1  \
1       SPD Smith  Maharashtra Cricket Association Stadium  A Nand Kishore   
2         CA Lynn   Saurashtra Cricket Association Stadium     Nitin Menon   
3      GJ Maxwell                   Holkar Cricket Stadium    AK Chaudhary   
6          N Rana                         Wankhede Stadium     Nitin Menon   
8       SV Samson  Maharashtra Cricket Association Stadium     AY Dandekar   

         umpire2 umpire3  match_id  total_runs  
1         S Ravi     NaN         2         184  
2      CK Nandan     NaN         3         183  
3  C Shamshuddin     NaN         4         163  
6      CK Nandan     NaN         7         178  
8         S Ravi     NaN         9         205